subiquity/models/filesystem.py

```diff
@@ -976,14 +1110,77 @@ def _probe_bootloader(self):
 
     def __init__(self):
         self.bootloader = self._probe_bootloader()
-        self._disk_info = []
+        self._probe_data = None
         self.reset()
 
     def reset(self):
-        self._actions = [
-            Disk.from_info(self, info) for info in self._disk_info]
+        if self._probe_data is not None:
+            config = storage_config.extract_storage_config(self._probe_data)
+            self._actions = self._actions_from_config(
+                config["storage"]["config"],
+                self._probe_data['blockdev'])
+        else:
+            self._actions = []
         self.grub_install_device = None
 
+    def _actions_from_config(self, config, blockdevs):
+        byid = {}
+        objs = []
+        exclusions = set()
+        for action in config:
+            if action['type'] == 'mount':
+                exclusions.add(byid[action['device']])
+                continue
+            c = _type_to_cls.get(action['type'], None)
+            if c is None:
+                # Ignore any action we do not know how to process yet
+                # (e.g. bcache)
+                continue
+            kw = {}
+            for f in attr.fields(c):
+                n = f.name
+                if n not in action:
+                    continue
+                v = action[n]
+                if f.metadata.get('ref', False):
+                    kw[n] = byid[v]
+                elif f.metadata.get('reflist', False):
+                    kw[n] = [byid[id] for id in v]
+                else:
+                    kw[n] = v
+            if kw['type'] == 'disk':
+                path = kw['path']
+                kw['info'] = StorageInfo({path: blockdevs[path]})
+            kw['preserve'] = True
+            obj = byid[action['id']] = c(m=self, **kw)
+            if action['type'] == "format":
+                obj.volume._original_fs = obj
+            objs.append(obj)
+
+        # We filter out anything that can be reached from a currently
+        # mounted device. The motivation here is only to exclude the
+        # media subiquity is mounted from, so this might be a bit
+        # excessive but hey it works.
+        while True:
+            log.debug("exclusions %s", {e.id for e in exclusions})
+            next_exclusions = exclusions.copy()
+            for e in exclusions:
+                next_exclusions.update(itertools.chain(
+                    dependencies(e), reverse_dependencies(e)))
+            if len(exclusions) == len(next_exclusions):
+                break
+            exclusions = next_exclusions
+
+        objs = [o for o in objs if o not in exclusions]
```

Comment:
if objs was a set, then this could be:

objs = objs.difference(exclusions)

Reply:
But (and this is a magnificent example of implicit tight coupling) the order returned actually matters! The devices are presented in the filesystem view in the reverse of the order they appear in _actions, which means that e.g. a RAID appears higher up the list than the disks is is composed of. I think this is quite important as it makes "unpeeling" existing compound structures easy, you just delete the top device until you only have disks left. I've added some comments about this.